### ISSUE TITLE
A way to pass a timeout limit to kpro:connect_partition_leader

### ISSUE THREAD

**Issue by escobera** [kafka4beam/brod#389]
Hello there,

I was having a timeout problem on the `brod_utils:resolve_offset/5` function that I tracked down all the way to https://github.com/escobera/kafka_protocol/blob/master/src/kpro_brokers.erl#L79 .

By increasing the default timeout to 30 seconds my problem was solved, but I couldn't find a way to pass this timeout from brod's api. Is there something I'm missing?

**Comment by zmstone** [comment]
one workaround is to call `kpro: connect_partition_leader/5` which supports the timeout to get the connection pid, then call `brod_utils:resolve_offset(pid(), topic(), partition(), offset_time())` with the leader connection pid.

We need a fix to support timeout option in the `/5` api. Or maybe reuse the connection timeout in `conn_config`

**Comment by escobera** [comment]
Ok, so maybe we can change 
https://github.com/escobera/kafka_protocol/blob/master/src/kpro_brokers.erl#L80 to
`Timeout = maps:get(connect_timeout, Config, ?DEFAULT_TIMEOUT),`

or change https://github.com/escobera/kafka_protocol/blob/master/src/kpro.erl#L368-L370 to

```
Timeout = maps:get(connect_timeout, ConnConfig, ?DEFAULT_TIMEOUT),
connect_partition_leader(Bootstrap, ConnConfig, Topic, Partition, #{timeout => Timeout}).
```

The second change seems less prone to mess with existing uses

**Comment by escobera** [comment]
What do you think @zmstone ?

**Comment by zmstone** [comment]
Hi. Sorry for the late reply.

I actually would prefer to have it fixed in brod instead.
i.e. add a `resolve_offset/6`
```
resolve_offset(Hosts, Topic, Partition, Time, ConnCfg) ->
  Opts = #{timeout => maps:get(connect_timeout, ConnConfig, ?DEFAULT_TIMEOUT)},
  resolve_offset(Hosts, Topic, Partition, Time, ConnCfg, Opts).

resolve_offset(Hosts, Topic, Partition, Time, ConnCfg, Opts) ->
   with_conn(
    kpro:connect_partition_leader(Hosts, ConnCfg, Topic, Partition Opts),
    fun(Pid) -> resolve_offset(Pid, Topic, Partition, Time) end).
```

**Comment by escobera** [comment]
I'll try that later tonight and send a PR your way.

**Comment by escobera** [comment]
https://github.com/klarna/brod/pull/393

**Comment by mikpe** [comment]
Fixed by #393
